### ISSUE TITLE
Mikro orm config error key is null

### ISSUE THREAD

**Issue by AndonMitev** [mikro-orm/mikro-orm#866]
**Describe the bug**
I'm trying to setup mikro orm and connect to postgresql but following error pop up:

TypeError [ERR_INVALID_ARG_TYPE]: The "key" argument must be of type string or an instance of Buffer, TypedArray, DataView, or KeyObject. Received null

**Stack trace**
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity Post
[discovery] - entity discovery finished, found 1 entities, took 12 ms
internal/crypto/keys.js:304
    throw new ERR_INVALID_ARG_TYPE(
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "key" argument must be of type string or an instance of Buffer, TypedArray, DataView, or KeyObject. Received null
    at prepareSecretKey (internal/crypto/keys.js:304:11)
    at new Hmac (internal/crypto/hash.js:113:9)
    at Object.createHmac (crypto.js:143:10)
    at createHMAC (C:\Users\earth\Desktop\FullStack\node_modules\pg\lib\sasl.js:133:17)
    at Hi (C:\Users\earth\Desktop\FullStack\node_modules\pg\lib\sasl.js:137:13)
    at Object.continueSession (C:\Users\earth\Desktop\FullStack\node_modules\pg\lib\sasl.js:32:24)
    at Client._handleAuthSASLContinue (C:\Users\earth\Desktop\FullStack\node_modules\pg\lib\client.js:248:10)
    at Connection.emit (events.js:315:20)
    at Connection.EventEmitter.emit (domain.js:483:12)
    at C:\Users\earth\Desktop\FullStack\node_modules\pg\lib\connection.js:109:12 {
  code: 'ERR_INVALID_ARG_TYPE'
}
[nodemon] app crashed - waiting for file changes before starting...

**Mikro ORM Config**

`
export default {
  migrations: {
    path: path.join(__dirname, "./migrations"),
    pattern: /^[\w-]+\d+\.[tj]s$/,
  },
  entities: [Post],
  dbName: "new-stuff",
  type: "postgresql",
  debug: !__prod__,
} as Parameters<typeof MikroORM.init>[0];
`

**Comment by B4nan** [comment]
Not sure what is wrong, but this has nothing to do with MikroORM, the error is thrown from `pg`, most probably you have the connection misconfigured (or your postgres server, hard to say). Maybe you need to provide password or something like that?

**Comment by B4nan** [comment]
Looking at the source code of `pg`, and the `createHMAC` is called with the password.

What versions are you using? How does your package.json look like? Are you able to connect to your database with 3rd party tool (like pg-admin, phpmyadmin, adminer, etc...)?

**Comment by AndonMitev** [comment]
![image](https://user-images.githubusercontent.com/32274987/94188356-f0275f00-feb1-11ea-96f1-09b9b043b077.png)
Yes i can see the db in 3rd party tool. I guess might be something with the password then! Thanks a lot

**Comment by B4nan** [comment]
How does the full config look like? Maybe you are using wrong fields (e.g. `hostname` instead of `host`, or `username` instead of `user`), the interface looks like this:

```ts
export interface ConnectionOptions {
  dbName?: string;
  name?: string;
  clientUrl?: string;
  host?: string;
  port?: number;
  user?: string;
  password?: string;
  charset?: string;
  multipleStatements?: boolean; // for mysql driver
  pool?: PoolConfig;
}
```

Doing `export default { ... } as Parameters<typeof MikroORM.init>[0]` is also not ideal, as it won't type check things correctly. Better to do it this way:

```ts
import { Options } from '@mikro-orm/core'; // `Options` is what you want, no need to get the type from init method

const config: Options = { ... };

export default config;
```

**Comment by AndonMitev** [comment]
Well this is how my config looks like:

`
const config: Options = {
  migrations: {
    path: path.join(__dirname, "./migrations"),
    pattern: /^[\w-]+\d+\.[tj]s$/,
  },
  entities: [Post],
  dbName: "postgres",
  type: "postgresql",
  debug: !__prod__,
};
`

**Comment by B4nan** [comment]
Can you share (for future readers) what was the issue? I guess you were missing a password in your config, and it was required by your postgres server?

**Comment by abdallahakrab** [comment]
not OP, had the exact same error when I omitted "user" from config, As mentioned it has nothing to do with Micro-orm

**Comment by sleepybyte23** [comment]
export default {
    migrations: {
        path: path.join(__dirname, "./migrations"), // path to the folder with migrations
        pattern: /^[\w-]+\d+\.[tj]s$/, // regex pattern for the migration files       
      },
        entities: [Post],
        dbName : 'lireddit',
        type: 'postgresql',
        password: "s3xy",
        debug: !__prod__        
  } as Parameters<typeof MikroORM.init>[0];


worked for me, I was missing password.

**Comment by Omar-Belghaouti** [comment]
@sleepybyte23 but when I tried to put the password as a env variable by using process.env.PASSWORD! for example I get the same issue

**Comment by Langstra** [comment]
Unsure what you are exactly trying, without any context it is hard to tell. If you are trying environment variables please check the documentation. https://mikro-orm.io/docs/configuration/#using-environment-variables

For other questions you could use the discussion feature on GitHub, try the mikroorm slack or stackoverflow. GitHub issues are for bugs, improvements or feature requests, not for support. If you truly believe you found a bug, then open a new issue with a full bug report, issue template and reproduction.

**Comment by Omar-Belghaouti** [comment]
> Unsure what you are exactly trying, without any context it is hard to tell. If you are trying environment variables please check the documentation. https://mikro-orm.io/docs/configuration/#using-environment-variables
> 
> For other questions you could use the discussion feature on GitHub, try the mikroorm slack or stackoverflow. GitHub issues are for bugs, improvements or feature requests, not for support. If you truly believe you found a bug, then open a new issue with a full bug report, issue template and reproduction.

@Langstra I tried to put the password in `.env` file like this:
```
PASSWORD=test
```
And use it in the config file like this:
```js
import { Options } from '@mikro-orm/core';

const config: Options = {
  migrations: {
    path: path.join(__dirname, "./migrations"), // path to the folder with migrations
    pattern: /^[\w-]+\d+\.[tj]s$/, // regex pattern for the migration files       
  },
  entities: [Post],
  dbName : 'lireddit',
  type: 'postgresql',
  password: process.env.PASSWORD,
  debug: !__prod__,
};

export default config;
```

And I get this:
```
error: sasl: scram-server-first-message: client password must be a string
```

**Comment by B4nan** [comment]
You need to initialize `dotenv` in that config file if you want to use it like this (the ORM does this internally, but at a later stage, after the CLI config is loaded). Or just use the ORM env var, which is `MIKRO_ORM_PASSWORD`.

**Comment by Omar-Belghaouti** [comment]
> You need to initialize `dotenv` in that config file if you want to use it like this (the ORM does this internally, but at a later stage, after the CLI config is loaded). Or just use the ORM env var, which is `MIKRO_ORM_PASSWORD`.

Thanks for the info
